src/main/java/mate/academy/service/impl/AuthenticationServiceImpl.java

```diff
@@ -0,0 +1,43 @@
+package mate.academy.service.impl;
+
+import java.util.Optional;
+import mate.academy.exception.AuthenticationException;
+import mate.academy.exception.RegistrationException;
+import mate.academy.lib.Inject;
+import mate.academy.lib.Service;
+import mate.academy.model.User;
+import mate.academy.service.AuthenticationService;
+import mate.academy.service.UserService;
+import mate.academy.util.HashUtil;
+
+@Service
+public class AuthenticationServiceImpl implements AuthenticationService {
+    private static final int SALT_SIZE = 16;
+    @Inject
+    private UserService userService;
+
+    @Override
+    public User login(String email, String password) throws AuthenticationException {
+        Optional<User> user = userService.findByEmail(email);
+        if (user.isPresent() && checkPassword(user.get(), password)) {
+            return user.get();
+        }
+        throw new AuthenticationException("Email or password was incorrect");
+    }
+
+    @Override
+    public User register(String email, String password) throws RegistrationException {
+        if (userService.findByEmail(email).isPresent()) {
+            throw new RegistrationException("Email " + email + "is already used");
+        }
+        User user = new User();
+        user.setEmail(email);
+        user.setSalt(HashUtil.getSalt(SALT_SIZE));
+        user.setPassword(HashUtil.hashPassword(password, user.getSalt()));
```

Comment:
Lets move this logic to UserService add() method, so we cant add user without salt and hashed password

Reply:
Should i change params of UserService.add? Now it's add(User user). If i keep like that - user.password when we call add - user input string, when we try to login - hashed string. I think it's not good.
So i can change UserService.add signature to add(User user, String password). Ok?

Reply:
No, you should not change add method signature, It should look like
```
    @Override
    public User add(User user) {
        user.setSalt(HashUtil.getSalt());
        user.setPassword(HashUtil.hashPassword(user.getPassword(), user.getSalt()));
        return userDao.add(user);
    }
```